Telechat: Assembly Upstreaming III

lukeg101 opened the pull request:
This patch is the last major set of assembly changes for herd, In detail it adds:
- Inline barrel shifter support for LDRB/H and STRB/H instructions
- Added TBZ and TBNZ instructions to AArch64
- Added __(u)int128_t C type, no unittests (see commit message for details)
- Added STXP instruction to AArch64
- Added LDXP instruction to AArch64
- Added LDAXP and STLXP instructions to AArch64
- Updated LDAR semantics to handle numeric addresses in AArch64
- Updated STXR semantics to handle numeric addresses in AArch64
- Updated label support for BNE,BEQ,CBZ,CBNZ,B,and BL in AArch64
- Merged this branch with upstream fork to make pull request proceed without the need for a manual merge

Unit tests are provided unless otherwise specified.

There may be other minor assembly changes later down the line, but for now this covers most of them.

lukeg101 commented:
For clarity new commits are from `6e75bd9` onwards

maranget commented on herd/AArch64Sem.ml:
```diff
@@ -428,6 +428,39 @@ module Make
+            >>= fun resa2 -> write sz ea2 v2 resa2 ii)
+
+        >>! B.Next
+
```
Here we are, something is wrong; STXP tests the contents of  the reservation register twice (have a look at write_mem_atomic, which is called for addr1 and addr2. However, ldar writes a into the register. As a result we (may ?) get an equation a = incr a, which has no solution.
I'd suggest to store a in the reservation register, as (ldar now does) and to test the contents of the reservation register  against the base address only in stxp.

I also observe that no unit test tests a successful LDXP/STXP pair.

lukeg101 replied:
I am going to revert the patch for LDXP too until we have vectors - I won't address this feedback until I resubmit after vectors

maranget commented on herd/unittests/AArch64/A29.litmus:
```diff
@@ -0,0 +1,10 @@
+  STXR W3, X0, [X2];
+
+(* We expect the instruction to fail without an accompanying load *)
+exists (z=0  /\ a=0 /\ 0:X3=1)
```
Although not very important, the comment suggests to replace exists by forall in the condition.

lukeg101 replied:
fixed

maranget commented on herd/unittests/AArch64/A43.litmus:
```diff
@@ -0,0 +1,11 @@
+  ADD X0, X0, #1;
+  STXR W3, X0, [X4];
+
+exists ((z=2 /\ 0:X3=1) \/ (z=3 /\ 0:X3=0))
```
Again `forall` may be more appropriate.

lukeg101 replied:
fixed

lukeg101 commented on gen/AArch64Compile_gen.ml:
```diff
@@ -248,8 +248,8 @@ module Make(Cfg:Config) : XXXCompile_gen.S =
     let stlr_mixed_idx sz r1 r2 idx  =
```
fair

lukeg101 commented on herd/AArch64Sem.ml:
```diff
@@ -537,6 +537,19 @@ module Make
                   >>= fun () -> B.bccT v l
+        | I_TBZ(_,r,k,l) ->
+            (read_reg_ord r ii)
+              >>= M.op Op.BitAt (V.intToV k)
```
fair, I will remove `BitAt` entirely and use `ReadBit` for `I_TBZ` and `I_TBNZ`

lukeg101 commented on litmus/AArch64Compile_litmus.ml:
```diff
@@ -123,6 +123,17 @@ let pp_shifter = function
+          memo k (A.Out.dump_label (tr_lab lbl)) in
+      { empty_ins with
+        memo; inputs=[r;]; outputs=[];
+        branch=[Next; Branch lbl] ; }
```
added

lukeg101 commented on lib/machSize.ml:
```diff
@@ -14,32 +14,36 @@
   | Short -> "short"
   | Word -> "word"
   | Quad -> "quad"
+  | QuadWord -> "quadword"
```
I am going to revert the 128 bit patch and implement vectors, then we can use a Vec 2 Int64 to represent 128 bit types

lukeg101 commented on herd/event.ml:
```diff
@@ -1251,6 +1261,98 @@ let do_para_comp es1 es2 =
+              then in_wresult else EventSet.empty));
+         EventRel.cartesian out_data1
+           (if C.variant Variant.Success || not (C.variant Variant.FullScDepend) then
+              EventSet.union in_wmem1 in_wmem2 else
+           EventSet.union3 in_wresult in_wmem1 in_wmem2);
```
This patch is also old, so I'm not sure either

lukeg101 commented on herd/unittests/AArch64/A44.litmus:
```diff
@@ -0,0 +1,11 @@
+  ADD X0, X0, #1      | ;
+  STXR W3, X0, [X4]   | ;
+
+exists ((not (z=3  /\ 0:X3=1)) \/ (not (z=6 /\ 0:X3=1)))
```
fixed

lukeg101 commented:
I have addressed all non-`ld*p`/`st*p`/`__(u)int128_t` issues, and reverted all `st*p`/`ld*p`,`__(u)int128_t` patches. 

I will follow up this code review with vector support for size bounded vectors in herd, which we can then use as syntactic sugar for `__(u)int128_t` types, and handle ld/st pairs in a uniform fashion for herd and litmus

lukeg101 commented:
As such all controversial changes are removed from this patch, so unless there is any other issues I think this is ready to merge.

As a side note, I have kept ldp/stp/int128 as reverts in the commit history so that it is easy to get the patches later and adapt them to vectors

maranget commented:
This is fine by me, let us merge.
